Added option to disable PageAction icon

mhweaver opened the pull request:
[As requested](http://www.reddit.com/r/companion/comments/j49v3/option_to_hide_the_icon_that_appears_on_the/), this adds an option to let you enable/disable the icon in the address bar.

mhweaver committed:
Added option to disable pageaction icon

chromakode commented on src/background.js:
```diff
@@ -525,17 +537,21 @@ window.addEventListener('storage', function(e) {
 initOptions()
 console.log('Shine loaded.')
 redditInfo.init()
+setAllPageActionIcons()
```
I would suggest checking `localStorage['showPageAction']` and skipping if it is false here, since there would be no need to loop over the tabs.

mhweaver replied:
I initially was thinking the same thing, but if I put it here, then `setPageActionIcon()` still gets called in several other places. If I put the check anywhere else, then it has to check the setting every time `setPageActionIcon` is called.

chromakode replied:
Ah, let me clarify: I was suggesting adding a check here _in addition_ to save a little effort in this special case.

mhweaver replied:
Gotcha

chromakode commented on src/background.js:
```diff
@@ -474,6 +475,14 @@ function onActionClicked(tab) {
   })
 }
 
+function handleOptionChange(id, value) {
```
There is an existing option change handler on [line 529](https://github.com/mhweaver/reddit-companion/blob/a2d6a44e7e7bc5c42c446d53df5c28b6cca048c2/src/background.js#L529) in the form of a `storage` event handler. The `storage` event gets fired whenever a localStorage value changes, such as when they are set in options.js. Can you please add this logic to a case in the `storage` event handler?

mhweaver replied:
Ah, didn't see that. Good to know

chromakode commented on src/background.js:
```diff
@@ -434,7 +435,7 @@ mailChecker = {
 
 function setPageActionIcon(tab) {
-  if (/^http:\/\/.*/.test(tab.url)) {
+  if (/^http:\/\/.*/.test(tab.url) && localStorage['showPageAction'] == 'true') {
```
Small optimization: can you please put the `localStorage['showPageAction'] == 'true'` condition first? That way, if the page action is disabled, the if statement will quit early rather than evaluating the regular expression. See [short circuit evaluation](http://en.wikipedia.org/wiki/Short-circuit_evaluation).

mhweaver replied:
Turns out how I have it set up here pretty much totally just disables the bar (whoops!), so I'm nesting it and having the check only actually affect whether or not the show/hide happens.

chromakode commented on src/background.js:
```diff
@@ -525,17 +537,21 @@ window.addEventListener('storage', function(e) {
+  chrome.windows.getAll({populate:true}, function(wins) {
+    wins.forEach(function(win) {
+      win.tabs.forEach(function(tab) {
+          chrome.pageAction.hide(tab.id)
```
I would suggest putting the `chrome.pageAction.hide` behavior inside `setPageActionIcon` so that `setPageActionIcon` does all the work for setting a page action icon to the proper state.

mhweaver committed:
Moved chrome.pageAction hide/show logic to setPageActionIcon()

Moved the check for the setting to happen along with the regular
setPageActionIcon stuff, since where it was was actually just disabling the bar
along with the icon (whoops).

Added option check for the initial setAllPageActionIcons() call that happens
when the extension loads.

Removed optionChangeHandler and moved logic to existing 'storage' event handler.

mhweaver commented:
Done. I did all the suggestions (good catches, btw), except for the optimization in `setPageActionIcon()`, since having the check right there was a really bad idea to begin with.

chromakode commented:
Looks great, thanks! Will merge up. :-)

mhweaver committed:
Changed setPageActionIcon() to only setup the icon when it will be shown

Also changed setPageActionIcon() so info is passed as a parameter

Having setPageActionIcon() be the place where info is set is counter-intuitive
and somewhat confusing. Passing it as a param makes more sense and makes the
code easier to follow.

mhweaver commented:
The way I had it, it was still calling `chrome.pageAction.setIcon` even when showPageAction was off, so I moved that inside the if-statement where it checks the option. After doing so, the function was getting pretty clunky and less readable (ifs within ifs within ifs! Oh, god! Where does it end!?), so I made `info` a parameter.

chromakode commented:
Makes sense. Thanks!

chromakode commented:
Merged in 82e6607c82. Thanks!
